neuralprophet/time_net.py

```diff
@@ -767,6 +830,165 @@ def compute_components(self, inputs):
                 )
         return components
 
+    def set_compute_components(self, compute_components_flag):
+        self.compute_components_flag = compute_components_flag
+
+    def loss_func(self, inputs, predicted, targets):
+        loss = None
+        # Compute loss. no reduction.
+        loss = self.config_train.loss_func(predicted, targets)
+        # Weigh newer samples more.
+        loss = loss * self._get_time_based_sample_weight(t=inputs["time"])
+        loss = loss.sum(dim=2).mean()
+        # Regularize.
+        steps_per_epoch = math.ceil(self.trainer.estimated_stepping_batches / self.trainer.max_epochs)
+        progress_in_epoch = 1 - ((steps_per_epoch * (self.current_epoch + 1) - self.global_step) / steps_per_epoch)
+        loss, reg_loss = self._add_batch_regularizations(loss, self.current_epoch, progress_in_epoch)
+        return loss, reg_loss
+
+    def training_step(self, batch, batch_idx):
+        inputs, targets, _ = batch
+        # Run forward calculation
+        predicted = self.forward(inputs)
+        # Calculate loss
+        loss, reg_loss = self.loss_func(inputs, predicted, targets)
+        # Metrics
+        if not self.minimal:
+            predicted_denorm = self.denormalize(predicted[:, :, 0])
+            target_denorm = self.denormalize(targets.squeeze(dim=2))
+            self.log_dict(self.metrics_train(predicted_denorm, target_denorm), **self.log_args)
+            self.log("Loss", loss, **self.log_args)
+            self.log("RegLoss", reg_loss, **self.log_args)
+        return loss
```

Comment:
Info: The `_step` functions are used by Lightning to execute the training procedure. After the `training_step`, Lightning automatically calls the optimizer and learning rate scheduler to update the model parameters.